src/main/java/org/candlepin/model/EnvironmentCurator.java

```diff
@@ -308,4 +311,121 @@ public int removeEnvironmentContentReferences(Owner owner, Collection<String> co
         return count;
     }
 
+    /**
+     * Determines if any of the provided {@link Environment} IDs does not exist or does not belong to the
+     * provided {@link Owner}.
+     *
+     * @param envIds
+     *  the environment IDs to check
+     *
+     * @param owner
+     *  the owner of the environment IDs
+     *
+     * @return all of the environment IDs from the provided list that do not exist or do not belong to the
+     *  provided owner
+     */
+    public Set<String> getNonExistentEnvironmentIds(Collection<String> envIds, Owner owner) {
+        if (envIds == null || envIds.isEmpty() || owner == null) {
+            return new HashSet<>();
+        }
+
+        String jpql = "SELECT DISTINCT env.id FROM Environment env " +
+            "WHERE env.id IN (:envIds) AND env.ownerId = :ownerId";
+
+        Set<String> distinctEnvIds = new HashSet<>(envIds);
+
+        Query query = this.getEntityManager()
+            .createQuery(jpql, String.class);
+
+        Set<String> actualEnvIds = new HashSet<>();
+        for (List<String> block : partition(distinctEnvIds)) {
+            List<String> result = query.setParameter("envIds", block)
+                .setParameter("ownerId", owner.getId())
+                .getResultList();
+
+            actualEnvIds.addAll(result);
+        }
+
+        // Remove all of the existing environment IDs to determine the environments that are unknown
+        distinctEnvIds.removeAll(actualEnvIds);
+
+        return distinctEnvIds;
+    }
+
+    /**
+     * Removes the {@link Consumer}s from all of the environments they currently exist in.
+     *
+     * @param consumerUuids
+     *  the UUIDs for all the consumers that should be removed from their environments
+     *
+     * @return the number of consumers removed from all environments
+     */
+    private int removeConsumersFromAllEnvironments(Collection<String> consumerUuids) {
+        if (consumerUuids == null || consumerUuids.isEmpty()) {
+            return 0;
+        }
+
+        String statement = "DELETE FROM cp_consumer_environments " +
```

Comment:
Yeah, ideally we can use JQPL or JPA criteria to remove the environment relationship to the consumers and not have to deal with any of the caching issues. However, I'm struggling to find a way to do so

Reply:
I couldn't find a way to handle this with JPA criteria or JPQL unfortunately. I removed the native queries to avoid the cache evictions and opted to load the Consumers in memory and update the `environmentIds` collection on the `Consumer`s objects.

Reply:
I need to go through and look at this again to see what our options are. One thing we were hitting with SCA certs is that the consumer lookup and whatever else Hibernate is doing was really costing us a lot in terms of performance. It got bad enough that dropping the second JPA call and going native only -- even with the cache evictions -- ended up being an order of magnitude faster.

Knowing what we know now wrt query spaces, query optimization around the IN operator, and Hibernate's L2 cache shenanigans, we should revisit this and perhaps profile our various approaches.

This also applies to the "return the objects that don't exist" queries in this PR. Something about them doesn't feel quite right to me, but I haven't really dedicated much time into thinking out a better approach.

Reply:
Yeah, we can probably improve the performance here by using native queries.

With the current approach, it takes 1.641 seconds to update the environments of 1000 consumers locally. However, this doesn't account for the production scale of the DB and I'm not sure we can test this without promoting the change to the stage environment and load testing the new endpoint. 

I'm not sure what our acceptable performance metrics are for the code review stage (before release and before hitting stage for loadtesting).